# Линейные модели и нейронные сети. Практическая работа

В этой практической работе пять обязательных задач.

Они помогут понять, что вы действительно усвоили материал модуля. 

Удачи!

## Цели практической работы
Потренироваться в обучении модели:
- линейной регрессии;
- логистической регрессии; 
- многослойного персептрона.

## Что входит в практическую работу

1. Загрузите датасет и ознакомьтесь с ним.
2. Обучите линейную регрессию, включая признаки с нулевым весом и без них. Замерьте качество.
3. Обучите случайный лес. Замерьте качество.
4. Обучите многослойный перцептрон. Замерьте качество.
5. Реализуйте стратегию голосования. Замерьте качество.

## Что оценивается

* Выполнены все задания. В каждом:
 * в коде нет ручных перечислений, все действия автоматизированы;
 * результаты вычислений и применённых операций корректны;
 * ответы на вопросы, где требуется, корректны и обоснованы; 
 * код читабелен: переменным даны осмысленные названия, соблюдены отступы и правила расстановки пробелов. Стилизация кода соответствует рекомендациям [PEP 8](https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html).
 
* Репозиторий проекта оформлен корректно:
 * содержит осмысленные коммиты, содержащие конкретные реализованные фичи;
 * ветки названы согласно назначению;
 * файлы, не связанные с проектом, не хранятся в репозитории.

## Как отправить работу на проверку

Сдайте практическую работу этого модуля через систему контроля версий Git сервиса Skillbox GitLab. После загрузки работы на проверку напишите об этом в личном кабинете своему куратору.

## Задача

Постройте модель классификации, определяющую категорию цены / цену подержанного автомобиля в зависимости от характеристик транспортного средства. 

Используйте датасет из коллекции подержанных автомобилей, выставленных на продажу в Соединенных Штатах. Он уже подготовлен, без выбросов и с категориальными фичами, преобразованными с помощью one hot encoding, и количественными фичами, стандартизированными с помощью скейлеров.

### Описание датасета:
- `id`— идентификатор записи;
- `is_manufacturer_name`— признак производителя автомобиля;

- `region_*`— регион;
- `x0_*`— тип топлива;
- `manufacturer_*`— производитель;
- `short_model_*`— сокращённая модель автомобиля;
- `title_status_*`— статус;
- `transmission_*`— коробка передач;
- `state_*`— штат;
- `age_category_*`— возрастная категория автомобиля;

- `std_scaled_odometer`— количество пройденных миль (после стандартизации);
- `year_std`— год выпуска (после стандартизации);
- `lat_std`— широта (после стандартизации);
- `long_std`— долгота (после стандартизации);
- `odometer/price_std`— отношение стоимости к пробегу автомобиля (после стандартизации);
- `desc_len_std`— количество символов в тексте объявления о продаже (после стандартизации);
- `model_in_desc_std`— количество наименований модели автомобиля в тексте объявления о продаже (после стандартизации);
- `model_len_std`— длина наименования автомобиля (после стандартизации);
- `model_word_count_std`— количество слов в наименовании автомобиля (после стандартизации);
- `month_std`— номер месяца размещения объявления о продаже автомобиля (после стандартизации);
- `dayofweek_std`— день недели размещения объявления о продаже автомобиля (после стандартизации);
- `diff_years_std`— количество лет между годом производства автомобиля и годом размещения объявления о продаже автомобиля (после стандартизации);
- `price`— стоимость;
- `price_category`— категория цены.

## Обязательные задачи

In [264]:
import matplotlib.pyplot as plt
import pandas as pd

from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.metrics import accuracy_score, confusion_matrix, mean_absolute_error
from sklearn.model_selection import train_test_split 
from sklearn.neural_network import MLPClassifier

In [265]:
df = pd.read_csv('data/vehicles_dataset_prepared.csv')
df.head()

id  price price_category  is_audi  is_ford  is_chevrolet  \
0  7308295377  54990           high        0        0             0   
1  7316380095  16942         medium        0        1             0   
2  7313733749  35590           high        0        0             0   
3  7308210929  14500         medium        0        0             0   
4  7303797340  14590         medium        0        0             0   

   is_toyota  x0_diesel  x0_electric  x0_gas  ...  long_std  year_std  \
0          0        1.0          0.0     0.0  ...  0.484245  1.322394   
1          0        0.0          0.0     0.0  ...  1.110800  0.695973   
2          0        0.0          0.0     1.0  ...  0.531185  0.852578   
3          1        0.0          0.0     1.0  ...  0.853562  0.226157   
4          0        0.0          0.0     0.0  ...  0.557607  0.069552   

   odometer/price_std  desc_len_std  model_in_desc_std  model_len_std  \
0           -0.510784      0.632075          -0.155788       1.163032   
1           -0.402947     -0.646781          -0.155788       0.932087   
2           -0.514480      0.560744          -0.155788       0.470197   
3           -0.241883      0.180435          -0.155788      -0.915473   
4           -0.333074      0.766366          -0.155788       1.163032   

   model_word_count_std  month_std  dayofweek_std  diff_years_std  
0              1.910669  -0.615846       1.120284       -1.322394  
1              1.235799   1.623784      -1.374972       -0.695973  
2              0.560930  -0.615846      -0.376870       -0.852578  
3             -0.788810  -0.615846       1.120284       -0.226157  
4              1.910669  -0.615846       0.122182       -0.069552  

[5 rows x 1463 columns]

**Задача 1. Линейная регрессия**

Вспомните задачу по предсказанию стоимости подержанного автомобиля. Попробуйте обучить модель линейной регрессии для предсказания цены автомобиля (колонка `price`). Для этого сделайте шаги:

- подготовьте данные: удалите колонки, которые косвенно содержат информацию о целевой переменной (`odometer/price_std`, `price_category`);
- разделите выборку на треин и тест в отношении 70/30;
- обучите модель линейной регрессии с дефолтными параметрами;
- посчитайте значение метрики mae на тестовой выборке для линейной регрессии;
- выведите получившиеся коэффициенты линейной регрессии при каждом параметре обучающей выборки с помощью метода `coef_` (есть ли коэффициенты, которые равны нулю? Если есть, выведите названия фичей с нулевым коэффициентом);
- удалите фичи, коэффициенты которых равны нулю; переобучите модель; убедитесь, что значение метрики не изменилось.



In [268]:
df_prepared = df.copy()
df_prepared = df_prepared.drop(['price_category', 'odometer/price_std', 'id'], axis=1)

x = df_prepared.drop(['price'], axis=1)
y = df_prepared['price']

x_train, x_test, y_train, y_test = train_test_split(
    x, y, test_size=0.3, random_state=42)

In [269]:
linear_reg = LinearRegression()
linear_reg.fit(x_train, y_train)
pred = linear_reg.predict(x_test)

mae = mean_absolute_error(y_test, pred)
print("Mean Absolute Error:", mae)

Mean Absolute Error: 10599856947839.219


In [270]:
coefficients = linear_reg.coef_
zero_coefs = []
for feature, coef in zip(x_train.columns, coefficients):
    print(f'{feature}: {coef}')
    if coef == 0:
        zero_coefs.append(feature)

is_audi: -55149865320786.414
is_ford: 6298661488619046.0
is_chevrolet: 2.7620866027500652e+16
is_toyota: 1875561093447261.2
x0_diesel: -500879554775124.2
x0_electric: -500879554783861.7
x0_gas: -500879554782460.25
x0_hybrid: -500879554781842.8
x0_other: -500879554781770.4
std_scaled_odometer: -3529.0
short_model_-benz: 37014806384181.97
short_model_1: 37014806382823.97
short_model_124: -1654550785792988.0
short_model_128i: 37014806381534.88
short_model_135i: 37014806385543.19
short_model_1500: 37014806389415.875
short_model_163869: 37014806390131.82
short_model_1928: 1167736385029272.2
short_model_1951: 37014806401445.01
short_model_1954: 37014806392388.27
short_model_1960s: 37014806386362.67
short_model_1966: 37014806431140.98
short_model_1968: 37014806412709.74
short_model_1972: 37014806401621.3
short_model_1977: 228198777962140.47
short_model_1978: 37014806395893.375
short_model_198.5: 37014806401217.12
short_model_1988: 37014806388748.71
short_model_1994: 37014806384740.93
short_mo

In [271]:
print("Features which have zero coefs",zero_coefs)

Features which have zero coefs ['manufacturer_ferrari']


In [272]:
df_prepared.drop(['manufacturer_ferrari'], inplace=True, axis=1)

In [273]:
pred = linear_reg.predict(x_test)

mae = mean_absolute_error(y_test, pred)
print("Mean Absolute Error:", mae)

Mean Absolute Error: 10599856947839.219


**Задача 2. Логистическая регрессия**

Теперь в рамках тех же данных попробуйте предсказать `price_category` с помощью алгоритма логистической регрессии. Предварительно из датафрейма удалите переменные, в которых косвенно содержится информация о целевой переменной (`odometer/price_std`, `price`). 

Для обученной модели:

- рассчитайте и выведите метрику качества (accuracy) на тренировочной выборке;
- сделайте предикт на тестовых данных и положите его в переменную `logreg_pred`;
- рассчитайте и выведите accuracy и confusion_matrix на тестовой выборке.

Обратите внимание, что это задание засчитывается, если: 
- accuracy на тренировочной выборке > 87%;
- accuracy на тестовой выборке > 75.5%.

Для улучшения метрики можно преобразовывать входные данные и тюнить параметры логистической регрессии.

In [285]:

df_prepared = df.copy()
df_prepared = df_prepared.drop(['price', 'odometer/price_std','id'], axis=1)

x = df_prepared.drop(['price_category'], axis=1)
y = df_prepared['price_category']

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)


In [286]:
logreg = LogisticRegression()
logreg.fit(x_train, y_train)

train_accuracy = accuracy_score(y_train, logreg.predict(x_train))
print("Accuracy on training set:", train_accuracy)

Accuracy on training set: 0.8551908510322294


C:\Users\tatev\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [287]:
logreg_pred = logreg.predict(x_test)

In [288]:
accuracy_score(y_test, logreg_pred)

0.7543312543312544

In [289]:
confusion_matrix(y_test, logreg_pred)

array([[801,  26, 170],
       [ 22, 767, 164],
       [156, 171, 609]], dtype=int64)

In [290]:
logreg.get_params()

{'C': 1.0,
 'class_weight': None,
 'dual': False,
 'fit_intercept': True,
 'intercept_scaling': 1,
 'l1_ratio': None,
 'max_iter': 100,
 'multi_class': 'auto',
 'n_jobs': None,
 'penalty': 'l2',
 'random_state': None,
 'solver': 'lbfgs',
 'tol': 0.0001,
 'verbose': 0,
 'warm_start': False}

In [366]:
lr_tuned = LogisticRegression(C=4, max_iter=700) 

In [367]:
lr_tuned.fit(x_train, y_train)

train_accuracy = accuracy_score(y_train, lr_tuned.predict(x_train))
print("Accuracy on training set:", train_accuracy)

Accuracy on training set: 0.8822218921728798


In [368]:
logreg_pred = lr_tuned.predict(x_test)

In [369]:
accuracy_score(y_test, logreg_pred)

0.7616077616077616

In [295]:
confusion_matrix(y_test, logreg_pred)

array([[803,  24, 170],
       [ 24, 769, 160],
       [150, 164, 622]], dtype=int64)

**Задача 3. Многослойный персептрон**

Решите задачу, поставленную в предыдущем задании, применив модель многослойного персептрона. С помощью данного алгоритма добейтесь лучших значений точности на тренировочной и тестовой выборках, чем на логистической регрессии. 

Увеличение точности в данном задании, по сравнению с предыдущим, должно быть больше увеличения 0.01 по метрике accuracy.

In [370]:
mlp = MLPClassifier(random_state=12, max_iter=600)
mlp.fit(x_train,y_train)

MLPClassifier(max_iter=600, random_state=12)

In [371]:
mlp_pred = mlp.predict(x_test)
accuracy_score(y_test,mlp_pred)

0.7685377685377686

In [372]:
train_accuracy = accuracy_score(y_train, mlp.predict(x_train))
print("Accuracy on training set:", train_accuracy)

Accuracy on training set: 1.0


In [373]:
confusion_matrix(y_test, mlp_pred)

array([[800,  32, 165],
       [ 20, 760, 173],
       [121, 157, 658]], dtype=int64)

In [374]:
confusion_matrix(y_test, logreg_pred)

array([[805,  25, 167],
       [ 22, 770, 161],
       [148, 165, 623]], dtype=int64)

**Задача 4. Сравнение с древовидными моделями**

Обучите модель случайного леса на тех же данных для предсказания `price_category`. Сравните качество с моделью логистической регрессии и многослойного персептрона. Словами опишите, какая из моделей в каких случаях работает лучше по результатам на тестовой выборке, обоснуйте свой выбор.

In [375]:
from sklearn.tree import DecisionTreeClassifier

In [383]:
clf = DecisionTreeClassifier()
clf.fit(x_train, y_train)

DecisionTreeClassifier()

In [384]:
clf_pred_train = clf.predict(x_train)
clf_pred = clf.predict(x_test)

In [385]:
print(accuracy_score(y_train, clf_pred_train))
print(accuracy_score(y_test, clf_pred))

1.0
0.6774081774081774


In [379]:
clf = DecisionTreeClassifier(max_depth=15, min_samples_split=15)
clf.fit(x_train, y_train)

DecisionTreeClassifier(max_depth=15, min_samples_split=15)

In [380]:
clf_pred_train = clf.predict(x_train)
clf_pred = clf.predict(x_test)

In [347]:
print(accuracy_score(y_train, clf_pred_train))
print(accuracy_score(y_test, clf_pred))

0.8523689291549087
0.6722106722106722


In [344]:
from sklearn.model_selection import GridSearchCV

# Define the parameter grid to search
param_grid = {
    'max_depth': [None, 10, 20, 30, 40, 50],  # List of max_depth values to try
    'min_samples_split': [2, 5, 10],  # List of min_samples_split values to try
    'min_samples_leaf': [1, 2, 4]  # List of min_samples_leaf values to try
}

# Instantiate Decision Tree Classifier
clf = DecisionTreeClassifier()

# Instantiate GridSearchCV
grid_search = GridSearchCV(clf, param_grid, cv=5)  # 5-fold cross-validation

# Perform grid search on the training data
grid_search.fit(x_train, y_train)

# Best parameters found
print("Best parameters found:", grid_search.best_params_)

# Evaluate performance on the test set
test_accuracy = grid_search.score(x_test, y_test)
print("Test set accuracy:", test_accuracy)

Best parameters found: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2}
Test set accuracy: 0.6722106722106722


Исходя из предоставленных результатов на тестовой выборке, можно сделать следующие выводы:

Логистическая регрессия:

Точность на тестовой выборке: 0.7602
Точность на обучающей выборке: 0.8858
Многослойный персептрон (MLP):

Точность на тестовой выборке: 0.7685
Точность на обучающей выборке: 1.0
Дерево решений:

Точность на тестовой выборке: 0.6729
Точность на обучающей выборке: 1.0
Анализ:

Логистическая регрессия демонстрирует хорошую точность как на обучающей, так и на тестовой выборках, что может свидетельствовать о том, что модель обладает хорошей обобщающей способностью и недостаточно переобучена.

Многослойный персептрон (MLP) показывает немного лучшие результаты по сравнению с логистической регрессией на тестовой выборке, при этом имея идеальную точность на обучающей выборке. Это может указывать на то, что модель смогла лучше улавливать сложные нелинейные зависимости в данных.

Дерево решений демонстрирует наименьшую точность на тестовой выборке, при этом имея также идеальную точность на обучающей выборке. Это может свидетельствовать о переобучении модели, которая может слишком сложно настраиваться на обучающие данные и плохо обобщать на новые данные.

Выводы:

Если нам важна интерпретируемость модели и мы предполагаем линейные зависимости в данных, то логистическая регрессия может быть хорошим выбором.

Если в данных присутствуют сложные нелинейные зависимости, то многослойный персептрон (MLP) может дать лучшие результаты, как это видно по результатам на тестовой выборке.

Дерево решений может быть менее предпочтительным выбором из-за его склонности к переобучению, особенно при отсутствии подходящего контроля за глубиной или другими параметрами модели.

**Задача 5. Стратегия голосования**

Реализуйте стратегию голосования для предсказания целевой переменной.
Голосование в задаче классификации — это когда несколько моделей выдают свои предикты, и финальным выбирается тот предикт, который предсказали большинство моделей.

Для реализации этой стратегии проделайте следующее:

- сохраните предсказания каждой из моделей (случайный лес, многослойный персептрон, логистическая регрессия) для тестовой выборки в датафрейм `pred_df`;
- в четвёртую колонку `target` положите тот класс, который предсказало большинство классификаторов; например, если в строке были значения `high, medium, medium`, в `target` нужно положить `medium`;

     если в строке три разных класса (`high, medium, low`), придумайте свою стратегию по выбору значения; самая простая стратегия — выбрать рандомно одно значение из трёх;

- посчитайте точность предсказания с помощью голосования; выведите значения метрик accuracy и confusion_matrix.

Добейтесь значения точности > 78%.

In [386]:
pred_df = x_test.copy()
pred_df['Logistic_Regression_Prediction'] = logreg_pred
pred_df['MLP_Prediction'] = mlp_pred
pred_df['Decision_Tree_Prediction'] = clf_pred

pred_df.head()

is_audi  is_ford  is_chevrolet  is_toyota  x0_diesel  x0_electric  \
6070        0        0             0          0        0.0          0.0   
8916        0        0             0          0        0.0          0.0   
4634        0        0             0          0        0.0          0.0   
9530        0        0             1          0        0.0          0.0   
8284        0        0             1          0        0.0          0.0   

      x0_gas  x0_hybrid  x0_other  std_scaled_odometer  ...  desc_len_std  \
6070     1.0        0.0       0.0            -0.565773  ...      1.463295   
8916     1.0        0.0       0.0            -0.512667  ...     -0.864776   
4634     1.0        0.0       0.0             1.629268  ...     -0.104523   
9530     1.0        0.0       0.0             0.539238  ...     -0.456081   
8284     1.0        0.0       0.0            -1.198239  ...     -0.954668   

      model_in_desc_std  model_len_std  model_word_count_std  month_std  \
6070          -0.155788      -0.684528              -0.78881  -0.615846   
8916          -0.155788       0.008307               0.56093  -0.615846   
4634           0.265739      -0.684528              -0.78881  -0.615846   
9530          -0.155788      -0.453583              -0.11394  -0.615846   
8284          -0.155788       0.008307              -0.11394  -0.615846   

      dayofweek_std  diff_years_std  Logistic_Regression_Prediction  \
6070      -1.374972       -0.852578                            high   
8916       0.122182        0.087054                          medium   
4634      -0.875921        2.279527                             low   
9530       0.621233       -0.226157                             low   
8284       1.120284       -1.009183                            high   

      MLP_Prediction  Decision_Tree_Prediction  
6070            high                      high  
8916          medium                    medium  
4634             low                       low  
9530             low                       low  
8284            high                    medium  

[5 rows x 1462 columns]

In [387]:
classifiers_predictions = pred_df.loc[:,['Logistic_Regression_Prediction','MLP_Prediction','Decision_Tree_Prediction']]

In [388]:
from collections import Counter

def majority_vote(predictions):
    combined_predictions = []
    for classifier_predictions in zip(*predictions):
        vote_counts = Counter(classifier_predictions)
        ## if 3 values differ will return the class label that occurs most frequently in the predictions
        majority_vote = vote_counts.most_common(1)[0][0]
        combined_predictions.append(majority_vote)
    return combined_predictions


combined_predictions = majority_vote([logreg_pred, clf_pred, mlp_pred])

pred_df['Vote_Strategy'] = combined_predictions

pred_df.head()

is_audi  is_ford  is_chevrolet  is_toyota  x0_diesel  x0_electric  \
6070        0        0             0          0        0.0          0.0   
8916        0        0             0          0        0.0          0.0   
4634        0        0             0          0        0.0          0.0   
9530        0        0             1          0        0.0          0.0   
8284        0        0             1          0        0.0          0.0   

      x0_gas  x0_hybrid  x0_other  std_scaled_odometer  ...  \
6070     1.0        0.0       0.0            -0.565773  ...   
8916     1.0        0.0       0.0            -0.512667  ...   
4634     1.0        0.0       0.0             1.629268  ...   
9530     1.0        0.0       0.0             0.539238  ...   
8284     1.0        0.0       0.0            -1.198239  ...   

      model_in_desc_std  model_len_std  model_word_count_std  month_std  \
6070          -0.155788      -0.684528              -0.78881  -0.615846   
8916          -0.155788       0.008307               0.56093  -0.615846   
4634           0.265739      -0.684528              -0.78881  -0.615846   
9530          -0.155788      -0.453583              -0.11394  -0.615846   
8284          -0.155788       0.008307              -0.11394  -0.615846   

      dayofweek_std  diff_years_std  Logistic_Regression_Prediction  \
6070      -1.374972       -0.852578                            high   
8916       0.122182        0.087054                          medium   
4634      -0.875921        2.279527                             low   
9530       0.621233       -0.226157                             low   
8284       1.120284       -1.009183                            high   

      MLP_Prediction  Decision_Tree_Prediction  Vote_Strategy  
6070            high                      high           high  
8916          medium                    medium         medium  
4634             low                       low            low  
9530             low                       low            low  
8284            high                    medium           high  

[5 rows x 1463 columns]

In [389]:
pred_df[pred_df['Logistic_Regression_Prediction'] != pred_df['MLP_Prediction'] ]

is_audi  is_ford  is_chevrolet  is_toyota  x0_diesel  x0_electric  \
1791        1        0             0          0        0.0          0.0   
623         0        0             0          0        0.0          0.0   
5441        0        1             0          0        0.0          0.0   
3014        0        0             0          0        0.0          0.0   
7920        0        0             0          0        0.0          0.0   
...       ...      ...           ...        ...        ...          ...   
3909        0        0             1          0        0.0          0.0   
3817        0        1             0          0        0.0          0.0   
8998        0        1             0          0        0.0          0.0   
7615        0        1             0          0        0.0          0.0   
585         0        1             0          0        0.0          0.0   

      x0_gas  x0_hybrid  x0_other  std_scaled_odometer  ...  \
1791     1.0        0.0       0.0             0.028379  ...   
623      1.0        0.0       0.0             0.092975  ...   
5441     1.0        0.0       0.0            -0.019818  ...   
3014     0.0        0.0       1.0            -1.339800  ...   
7920     1.0        0.0       0.0             1.957717  ...   
...      ...        ...       ...                  ...  ...   
3909     1.0        0.0       0.0             0.327917  ...   
3817     1.0        0.0       0.0            -0.417040  ...   
8998     1.0        0.0       0.0             1.033315  ...   
7615     1.0        0.0       0.0            -1.082609  ...   
585      1.0        0.0       0.0             0.609907  ...   

      model_in_desc_std  model_len_std  model_word_count_std  month_std  \
1791           0.265739      -1.146418              -0.78881  -0.615846   
623           -0.155788      -0.800000              -0.78881  -0.615846   
5441          -0.155788       0.932087              -0.11394   1.623784   
3014          -0.155788       0.701142               0.56093  -0.615846   
7920          -0.155788       0.354725              -0.11394  -0.615846   
...                 ...            ...                   ...        ...   
3909          -0.155788      -0.569055              -0.78881  -0.615846   
3817          -0.155788      -0.915473              -0.78881   1.623784   
8998          -0.155788      -0.915473              -0.78881   1.623784   
7615           0.265739       0.585670               0.56093   1.623784   
585           -0.155788      -0.453583              -0.78881   1.623784   

      dayofweek_std  diff_years_std  Logistic_Regression_Prediction  \
1791       0.621233       -0.226157                             low   
623       -0.376870        2.122922                             low   
5441      -1.374972       -0.069552                            high   
3014       0.122182       -1.322394                          medium   
7920       0.621233        0.087054                             low   
...             ...             ...                             ...   
3909      -0.875921        0.087054                          medium   
3817       1.120284        1.183290                             low   
8998       1.120284       -0.539367                            high   
7615       1.120284       -0.852578                            high   
585       -0.875921        0.243659                             low   

      MLP_Prediction  Decision_Tree_Prediction  Vote_Strategy  
1791          medium                       low            low  
623           medium                    medium         medium  
5441          medium                      high           high  
3014            high                    medium         medium  
7920          medium                       low            low  
...              ...                       ...            ...  
3909             low                       low            low  
3817          medium                       low            low  
8998          medi

In [390]:
accuracy_score(y_test,pred_df['Vote_Strategy'])

0.781011781011781

**Примечание**

В этой практической работе перед вами встал выбор: включать ли колонку `id`. При удалении данной колонки во время обучения логистической регрессии качество заметно улучшается.

Обучать любую модель (будь то логистическая, линейная регрессия или древовидный алгоритм) на данных айдишников не считается хорошей практикой. Как правило, между `id` и целевой переменной не должно быть никаких взаимосвязей. Включая колонку `id` в качества атрибута в обучение, вы стараетесь подогнать результаты своей модели под айдишники записей. Таким образом модель обучится на некотором наборе частных случаев и, возможно, не обратит внимание на общие зависимости.   